Assets/Mirror/Runtime/NetworkManager.cs

```diff
@@ -358,7 +358,15 @@ public void StopClient()
             ClientScene.DestroyAllClientObjects();
             if (!string.IsNullOrEmpty(offlineScene))
             {
-                ClientChangeScene(offlineScene, false);
+                if (dontDestroyOnLoad)
+                {
+                    // Move the NetworkManager from the virtual DontDestroyOnLoad scene to the Game scene.
+                    // This let's it be destroyed when client changes to the Offline scene.
+                    SceneManager.MoveGameObjectToScene(gameObject, SceneManager.GetActiveScene());
```

Comment:
this looks very game specific.
can you explain why this should be in all projects?

Reply:
It's not at all game specific unless you're saying having a Network Manager is.  That's the only thing that's being moved out of DDOL - the Network Manager (or whatever inherited class someone is using).

Reply:
what is DDOL?

Reply:
DDOL = Dont Destroy On Load

Reply:
That virtual Scene Unity creates.

Reply:
I don't fully understand why this is necessary.
the point of DontDestroyOnLoad is to never destroy it and keep it in there forever.
why not leave it in dontdestroyonload? what is the issue that we are trying to fix?

Reply:
This is only about the **Offline scene** -- this code *only* comes into play from StopClient.  The player is leaving the game and going back to the start of the app.  They're starting over.

Reply:
and why does it not work if we keep netmanager in dontdestroyonload?
when starting offline scene for the first time, it is put into dontdestroyonload anyway, isn't it?

Reply:
Examples:
The match is over - go back to the match maker in Offline
The player is a jerk and got kicked
The player's internet disconnected
The game has an Exit button that should send them back to Offline.

Reply:
yes, but why does it not work if we keep the network manager in dontdestroyonload?

Reply:
Because it may have been in the Offline scene to begin with, so keeping it gives us two and there can only be one.  This ensures the player gets a clean fresh start if there is one there.  If there isn't one there, then we should not bring one there on top of whatever the game dev has created in the Offline scene where the app starts, e.g. the match maker, login, etc.

Reply:
did you run into a real problem with that, or did you add the code without a problem?
right now, networkmanager will destroy itself when trying to initialize the singleton and it already exists.
for example, my network zones addon for uMMORPG also switches between scenes all the time, and the first one loads a new NetworkManager that immediately destroys itself because the original one is still in DontDestroyOnLoad - which works perfectly fine :)

Reply:
Yes, I had to cure this in Lobby, and Paradocks just brought this up in Discord, and this has come up for others before.  This isn't about scene switching from one game scene to another.  This is only about when NetworkManager has an offline scene assigned to that field in the inspector.  No other mechanics are effected by this.  As I recall you don't even use the Online / Offline scene inspector fields (correct me if I'm wrong).  Regardless, if the game dev does **not** have a NM in the Offline scene on purpose, we should **not** bring him one unexpectedly.  He has other stuff there and a NM doesn't belong there.

Reply:
If there's nothing assigned for the Offline scene field, this code isn't reached.

Reply:
'if the game dev does not have a NM in the Offline scene on purpose, we should not bring him one unexpectedly. He has other stuff there and a NM doesn't belong there.'
why would that be the case? the offline scene needs a networkmanager to call start/connect and make it go to the online scene, or not?

Reply:
No, an offline scene may not have anything to do with Mirror...it might be a matchmaking screen, for example, that uses some other service outside of Mirror.

Reply:
Weather they get a NetworkManager in the offline scene should be dependent on the DontDestroyOnLoad setting on the NetworkManager. Same logic should apply to the online/offline scene loading. If they are blank dont load them. At that point all the power is left with the developer to do with as they wish.

Reply:
They need the DontDestroyOnLoad  to carry it around in game play scenes...that's different than the player quitting the game.

Reply:
That depends on their design inside the Online/Offline scene. I am offering flexibility to do both based on not forcing them to design around Mirror.

Reply:
As I've said, if they DO have NM in Offline scene, our destroying the one that we're carrying around  is harmless (and IMHO better) because a new one will be created, and if they DON'T have NM in offline we should not shove one in there for them and force them to work around Mirror to get rid of it.

Reply:
With this change, they don't have to work around Mirror at all - the flexibility is there, the need to work around is *removed* by this PR.

Reply:
@MrGadget1024 can we do the clientchangescene(offlinescene,true) change first please?
moving the networkmanager out of dontdestroyonload is just a bit too much magic imho. the whole point of DontDestroyOnLoad is to not destroy it on load, but then we destroy it on load.

as far as I understand, UNET was meant to have an offline scene with networkmanager in it, which then changes to online scene when online - or back to offline scene when offline again.

Reply:
I will make the change, but I have doubts that's how people actually use the online / offline scene mechanics.